### DISCUSSION TITLE
judgement offset

### DISCUSSION THREAD

**Discussion by iSimpFor** [ppy/osu#16252] [Ideas]
the idea is that if you have high input latency, you should be able to shift the timing windows later to compensate for each input lagging behind

this can be used if there are offset issues but the music, visuals, and sounds sync properly

**Comment by bdach** [comment]
this is essentially the same as https://github.com/ppy/osu/discussions/13143, and therefore infeasible. at least not without disabling hitsounds.

consider this: you hear the sound correctly, you hit the button, the game registers it 100ms later. when should the hitsound be played? by the time the game has finally received the button press, it is *too late to play the hitsound*.

**Reply by iSimpFor** [reply]
if that is the problem, i think that the best idea be to automatically disable hitsounds when said setting is adjusted far enough(or to limit the setting to a small range, a warning/notice will need to be put either way)
all rulesets will be pretty affected, especially standard, but the game should still be reasonable to play even without hitsounds

otherwise, the only solution would be to fix the cause of the latency which may not be reasonable in certain circumstances

**Reply by peppy** [reply]
Easy solution is to turn down hitsounds.

**Comment by Walavouchey** [comment]
I'm highly for adding a judgement offset that mutes hitsounds.

The proposed solution by @peppy is to "turn down hitsounds" (and use universal offset to account for input latency). This has the following side-effects:

- Storyboards, videos, and storyboarded sound samples become out of sync with the music (non-issue when not playing with storyboards)
- Beatmap hit objects become out of sync with the music in the same way (not a problem for "low-enough" input latency, but unplayable when large UO adjustments are required)
- Muting hitsounds either entails muting all sound effects (often undesirable), or muting them by using silent audio files in skins

The proposed judgement offset (or whatever you want to call it), that solely accounts for input (and audio) latency at the cost of muting hitsounds, would solve *all* of the above.

Apparently UO is supposed to be used to sync beatmap, storyboards, videos and sound samples to the music, but I have no idea how they would become out of sync in the first place (as any audio latency would just delay all of them together). Audio-input sync issues are at least often more prevalent due to input and/or audio latency, and since UO is the only option, it gets used and side-effects have to be worked around. (Local offset in stable does exactly the same thing as UO but in reverse as far as I'm aware through testing.)

Not entirely sure why it's called "Audio offset" in lazer when it shifts everything but the music, although I suppose you could say the music (excluding other audio, again despite the name) does get "offset" relative to everything else.

**Reply by peppy** [reply]
osu! is a game where hot sounds dare a main focus. i dont think we will ever offer an option that disabled these out of the box. hit sound volume as a separate volume control if the closest thing, discussed elsewhere.

**Reply by Walavouchey** [reply]
If you're considering adding a volume control for hitsounds, then what is stopping the idea of adding a dedicated judgement offset that would turn that volume knob all the way down when used (maybe even with a warning)?

I'm sure you understand that many people listen to their keytaps when playing rhythm games to stay on time. Having to hit early to time delayed hitsounds to the music is akin to playing an instrument with *latency* between input and you hearing the sound. This trade-off between hitting on time and hit objects/storyboards/videos being off is simply a deterrent for people trying to enjoy the game just like many other rhythm games.

You either have to tap early, or tap on time and try not to get thrown off by the hit objects, video and storyboard being way behind the music. I refuse to disregard the work of storyboarders by turning off storyboards, but even if I did that, it would only fix one of the issues I've outlined above. A hitsound volume control would also fix the third point, but a dedicated judgement offset would fix all of them and reach the best compromise. In all cases, hitsounds are best turned off anyway because of not having excellent hardware, unless you're willing to play with noticeably delayed input.

**Reply by smoogipoo** [reply]
> as any audio latency would just delay all of them together

The beatmap doesn't get offset. The most common case is audio device buffering causing the track to start playing later than 0, so you'd need a universal offset to bring the beatmap and music in sync again.

I think it would be interesting to see how a judgement offset would work _without_ muting. A 5ms delay could be detrimental in terms of hit windows (300 window is 13ms @ OD11), but is practically inaudible.

I'm also having a hard time accepting the following:
1) Universal offset being anything but a positive number (i.e. audio can only be delayed, it can't happen earlier than 0). Unless there's a case I'm missing here like an audio device skipping audio...?
2) Input delay being anything but a positive number (i.e. input can only be delayed).

I think we probably do need to put more work into defining offsets and their specific use-cases better.

**Reply by Walavouchey** [reply]
> The beatmap doesn't get offset. 

UO makes everything except the music get offset in time (including the beatmap), which is evident by simply shifting UO during gameplay (see https://github.com/ppy/osu/discussions/16256, where although I haven't put circles on every tick, I've also tested that the beatmap is always synced to video/SB/SB samples, and not necessarily the music)

> as any audio latency would just delay all of them together

My bad, it seems I've explained it wrong. What I meant to say was that any latency from the game to an audio playback device causes all audio (i.e. music, samples and hitsounds) to be delayed together, which means that there is no point in using UO to sync music and samples because any cause for those two being out of sync would not be because of audio latency from a playback device. SB samples, just like the music, are non-reactive to input, so I have no clue how those even could get out of sync unless it's a software bug. That was the original point of https://github.com/ppy/osu/discussions/16256, when I was only considering the samples, until I realised UO also affects everything else.

> Universal offset being anything but a positive number (i.e. audio can only be delayed, it can't happen earlier than 0). Unless there's a case I'm missing here like an audio device skipping audio...?

A positive UO makes the beatmap, video, storyboard, storyboard sound samples and hitsounds become shifted *back*, so they play early relative to the music (again, see https://github.com/ppy/osu/discussions/16256). Now consider a positive input and audio delay. You want to delay the judgements in order to account for delay in input and audio (if your tapping is timed to the music; if you're timing hitsounds to the music, you're best using 0 UO because you're already tapping early to account for this). You achieve that through a negative UO, which shifts everything *forward* relative to the music.

Yes, UO makes the *music* play relatively later or earlier according to its sign, but the offset is not applied to the music. Music would otherwise be distorted or seeked around while adjusting UO during gameplay.

Local offset (in stable) is the exact opposite, where a positive offset shifts everything forward relative to the music, which might be what you're thinking of.

> I think it would be interesting to see how a judgement offset would work without muting. A 5ms delay could be detrimental in terms of hit windows (300 window is 13ms @ OD11), but is practically inaudible.

This would only be advantageous for people who actually have at most maybe 10 ms of input to audio latency. Any more and you'd have to mute hitsounds because they'd be too delayed by the offset.

I'd be curious to see what UO you get if you timed your inputs relative to the music (and not the hitsounds relative to the music). This corresponds to your input-audio latency on your machine, assuming accurately timed hits. I get this to be at around -67, which means I have about 67 ms of input-audio latency. I'll of course endeavour to lower that, but I'm limited by my hardware rn.

**Reply by peppy** [reply]
> My bad, it seems I've explained it wrong. What I meant to say was that any latency from the game to an audio playback device causes all audio (i.e. music, samples and hitsounds) to be delayed together, which means that there is no point in using UO to sync music and samples because any cause for those two being out of sync would not be because of audio latency from a playback device. SB samples, just like the music, are non-reactive to input, so I have no clue how those even could get out of sync unless it's a software bug. That was the original point of #16256, when I was only considering the samples, until I realised UO also affects everything else.

It's not that simple. iI'm not about to explain this again in great detail, but:

the music is a stream. it is the master clock. the samples are not. they are one-off invocations. streams and samples are played back differently. therefore there is often a latency between the two. 

also there's no difference in stable or lazer except he offset direction has been corrected to make more sense.

please wait for exclusive wasapi support. this endeavour and conversation is unfortunately quite pointless. if you disagree then you don't understand the underlying issues properly. which is fine. offsets are hard.

**Reply by Walavouchey** [reply]
> It's not that simple. I'm not about to explain this again in great detail, but:
> 
> the music is a stream. it is the master clock. the samples are not. they are one-off invocations. streams and samples are played back differently. therefore there is often a latency between the two.

Thank you for the explanation, but that was not my concern. I've accepted that UO was designed to account for that.

> this endeavour and conversation is unfortunately quite pointless.

I've already made my point. I was simply responding to smoogi's response and clarifying. Nothing more. I'm not and haven't been expecting anything out of this conversation since your first response, and will simply wait for if and when the issues will be addressed, in any way shape or form.

**Reply by peppy** [reply]
No worries. Just avoiding going through this whole thing again. It comes up at least 3 times a year and ends in everyone reaching the same conclusion at the end after figuring out what offsets are. Offsets are hard and that's why we are still only maintaining one setting for this for the time being. As I've already mentioned before, users are allowed/able to use this to cater to not just audio issues but also input and whatever else they are trying to fix.

I'm more interested in reducing latency rather than introducing more complexity to the offset settings.

**Reply by iSimpFor** [reply]
i would also like the latency to be reduced but chances are, you may need linux expertise to figure out how to reduce it for linux users
someone else can help though

i can't use the exe because i can't figure out how to run as a regular user and i would prefer to avoid using wine if there is a linux-native application
